src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+not a constant pattern:
+
+```dart
+// List or map pattern:
+case [a, b]: ...
+
+// List or map literal:
+case const [a, b]: ...
+```
+
+## Identifier	
+
+`foo, _`
+
+Identifier patterns may behave like a [constant pattern](#constant) or like a
+[variable pattern](#variable), depending on the context where it appears:
+
+- [Declaration][] context: declares a new variable with identifier name:
+  `var (a, b) = (1, 2);`
+- [Assignment][] context: assigns to existing variable with identifier name:
+  `(a, b) = (3, 4);`
+- [Matching][] context: treated as a named constant pattern (unless its name is `_`):
+  ```dart
+  const c = 1;
+  switch (2) {
+    case c: print('match $c');
+    default: print('no match');    // Prints "no match".
+   }
+  ``` 
+- [Wildcard](#wildcard) identifier in any context: matches any value and discards it:
+  `case [_, var y, _]: print('The middle element is $y');`
+
+## List
+
+`[subpattern1, subpattern2]`
+
+List patterns match values that implement [`List`][], and then recursively match
+its subpatterns against the list's elements to destructure them by position.
+
+List patterns require that the pattern match the entire list, but you can use
+[rest elements](#rest-element) as a place holder.
+
+### Rest element
+
+List patterns can contain _one_ rest element (`...`) which allows matching lists
+of arbitrary lengths.
+
+```dart
+var [a, b, ..., c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $c $d'); // Prints "1 2 6 7".
+```
+
+Rest elements can also have subpatterns that collects elements that don't match
+the other subpatterns in the list, into a new list:
+
+```dart
+var [a, b, ...rest, c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $rest $c $d'); // Prints "1 2 [3, 4, 5] 6 7".
+```
+
+## Logical-and	
+
+`subpattern1 && subpattern2`
+
+A pair of patterns separated by `&&` matches only if both subpatterns match. If the
+left branch does not match, the right branch is not evaluated. 
+
+## Logical-or
+
+`subpattern1 || subpattern2`
+
+A pair of patterns separated by `||` matches if either of the branches match.
+If the left branch matches, the right branch is not evaluated. 
+
+Logical-or patterns can be nested inside a destructuring pattern. 
+
+```dart
+switch (list) {
+  // Matches a two-element list whose first element is 'a' or 'b':
+  case ['a' || 'b', var c]:
+}
+```
+
+{% comment %}
+[TODO: move the below content to the switch page, doesn't belong here]
+{% endcomment %}
+
+Logical-or patterns are useful for having multiple cases share a body in switch
+expressions or statements.
+
+```dart
+var isPrimary = switch (color) {
+  Color.red || Color.yellow || Color.blue => true,
+  _ => false
+};
+```
+
+Switch statements can have multiple cases share a body without using logical-or
+patterns, but they are still uniquely useful for allowing multiple cases to share
+a guard: 
+
+```dart
+switch (shape) {
+  case Square(size: var s) || Circle(size: var s) when s > 0:
+    print('Non-empty symmetric shape');
+}
+```
```

Comment:
Maybe mention that or patterns may bind variables, but both sub-patterns must bind the same variables (as in your example)?

Reply:
Added:

> Subpatterns in a logical-or pattern can bind variables, but the two branches must
> define the same set of variables, because only one branch will be evaluated when
> the pattern matches.